tls/s2n_handshake_io.c

```diff
@@ -852,6 +852,20 @@ message_type_t s2n_conn_get_current_message_type(struct s2n_connection *conn)
     return ACTIVE_MESSAGE(conn);
 }
 
+static S2N_RESULT s2n_execute_handler(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* Ensure the state machine referenced is consistent throughout the handshake */
+    if (conn->actual_protocol_version_established || IS_HELLO_RETRY_HANDSHAKE(conn) || WITH_EARLY_DATA(conn)) {
+        RESULT_ENSURE_EQ(IS_TLS13_HANDSHAKE(conn), conn->is_tls13_state_machine);
+    }
```

Comment:
Isn't it the MIDDLEBOX_COMPAT and EARLY_CLIENT_CCS flags that get set before actual_protocol_version_established is true, rather than the early data flag itself? I think the early data flag is only ever set after actual_protocol_version is set.

Reply:
Darn I think you're right. I'm going to have to figure out another way of doing this.